llvm/lib/Support/APFloat.cpp

```diff
@@ -1727,6 +1752,8 @@ IEEEFloat::opStatus IEEEFloat::normalize(roundingMode rounding_mode,
   /* Canonicalize zeroes.  */
   if (omsb == 0) {
     category = fcZero;
+    if (!APFloat::hasZero(*semantics))
+      makeSmallestNormalized(false);
```

Comment:
Not sure why you need to do anything here, or why the category is fcZero if there is no zero

Reply:
This handles the case where the all-zero encoding is valid but just that it does not represent zero for some semantics (and represents the smallest-normal value instead). I have added a comment there explaining this case.